routers/repo/http.go

```diff
@@ -122,15 +127,81 @@ func HTTP(ctx *context.Context) {
 	// Only public pull don't need auth.
 	isPublicPull := !repo.IsPrivate && isPull
 	var (
-		askAuth      = !isPublicPull || setting.Service.RequireSignInView
-		authUser     *models.User
-		authUsername string
-		authPasswd   string
-		environ      []string
+		askAuth  = !isPublicPull || setting.Service.RequireSignInView
+		authUser *models.User
+		environ  []string
 	)
 
 	// check access
 	if askAuth {
+		if authUser = checkAuth(ctx); authUser == nil {
+			return
+		}
+
+		perm, err := models.GetUserRepoPermission(repo, authUser)
+		if err != nil {
+			ctx.NotFound("GetRepositoryName", err)
+			return
+		}
+
+		if !perm.CanAccess(accessMode, unitType) {
+			ctx.NotFound("GetRepositoryByName", nil)
+			log.Error("%s", "User permission denied")
+			return
+		}
+
+		if !isPull && repo.IsMirror {
+			ctx.HandleText(http.StatusForbidden, "mirror repository is read-only")
+			return
+		}
+
+		environ = []string{
+			models.EnvRepoUsername + "=" + username,
+			models.EnvRepoName + "=" + reponame,
+			models.EnvPusherName + "=" + authUser.Name,
+			models.EnvPusherID + fmt.Sprintf("=%d", authUser.ID),
+			models.ProtectedBranchRepoID + fmt.Sprintf("=%d", repo.ID),
+		}
+
+		if !authUser.KeepEmailPrivate {
+			environ = append(environ, models.EnvPusherEmail+"="+authUser.Email)
+		}
+
+		if isWiki {
+			environ = append(environ, models.EnvRepoIsWiki+"=true")
+		} else {
+			environ = append(environ, models.EnvRepoIsWiki+"=false")
+		}
+	}
+
+	HTTPBackend(ctx, &serviceConfig{
+		UploadPack:  true,
+		ReceivePack: true,
+		Env:         environ,
+	})(ctx.Resp, ctx.Req.Request)
+}
+
+func checkAuth(ctx *context.Context) (authUser *models.User) {
+	var (
+		authUsername string
+		authPasswd   string
+		err          error
+	)
+
+	authUsername = ctx.Req.Header.Get(setting.ReverseProxyAuthUser)
+	if setting.Service.EnableReverseProxyAuth && len(authUsername) > 0 {
+		authUser, err = models.GetUserByName(authUsername)
+		if err != nil {
+			ctx.HandleText(401, "reverse proxy login error, got error while running GetUserByName")
+			return
+		}
+	} else {
+		authHead := ctx.Req.Header.Get("Authorization")
+		if len(authHead) == 0 {
+			ctx.Resp.Header().Set("WWW-Authenticate", "Basic realm=\".\"")
```

Comment:
Very strange. This line shows up as brand new from your code, except it isn't! It's in the original `http.go` file. I should see a "diff minus" section with these lines if for example you moved them, or you changed the indentation. How can that be? I can see your branch is correctly based on the current master, so I'm puzzled.

![image](https://user-images.githubusercontent.com/18600385/64497781-47e33000-d287-11e9-8a6d-530878e261db.png)

That's what caught my attention about you "implementing" a lot of auth stuff.